Add support for unqualified USE statements (for 586 and 323)

arporter opened the pull request:
Alters the way USE statements are handled by putting more intelligence into the ContainerSymbol object. This now stores a list of the symbols which are imported from it into the currect scope. It also, separately, keeps a record of whether there is a "wildcard" import of all of its public symbols.
This then allows us to create the correct USE statements in the Fortran backend.

arporter commented:
Ready for first review, probably from either @sergisiso or @rupertford.

rupertford commented:
Back to @arporter.

arporter commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -239,26 +240,65 @@ class FortranWriter(PSyIRVisitor):
+                        dsym.interface.container_symbol.name))
+            only_list.append(dsym.name)
+        # Another consistency check
+        if not only_list and not symbol.has_wildcard_import:
```
I've googled but to no avail. Please could you remind me of the purpose of such a thing? (Is it something to do with keeping things in scope so that they don't get garbage collected?)

rupertford replied:
Yes, I think so. If a module variable goes out of scope without having the save attribute (or being initialised which implies save apparently) then its value is undefined. Having said that, some compilers save the values anyway but it should not be assumed that they will. If we are not going to support this then the front end should reject it, not the back end. Personally I think we should support it and add a note in the PSyIR documentation explaining what it means.

arporter replied:
Yes, I agree. I will "make it so"...

arporter replied:
In principle, it would be valid Fortran to have:

    module some_mod
      use my_mod, only:
      use my_mod
    end module some_mod

Since neither of these `USE`s have an associated list of symbols, in the current implementation we have no way of distinguishing that case from a single `use my_mod` (wildcard import). Practically speaking, if a piece of code does have both `USE my_mod, ONLY:` *and* `USE my_mod` then the latter effectively renders the former redundant. In fact, *any* import of data from `my_mod` renders the use of `USE my_mod, ONLY:` redundant. We therefore have two choices:

  1) Do not explicitly capture the fact that there is an ONLY with an empty list. This means that
  any other USEs of that module will result in the USE with an empty ONLY list being omitted
  from the generated Fortran. This is functionally correct.
  2) Explicitly capture the presence of an ONLY with an empty list. This means that the Fortran
  backend can generate syntactically identical code to that found by the frontend.

So far, we've always designed things such that no information is lost when converting from Fortran into the PSyIR. This favours option 2. However, I'm not certain this is the best approach here since all we're doing is retaining information that the original Fortran was non-sensical.

arporter replied:
After discussion in the office, we've decided that option 1. is best as there's no point polluting the PSyIR with redundancy.

arporter commented:
Assuming Travis is happy then this is ready for second review.

arporter commented:
That was a bit involved but it has forced me to improve things (ensuring consistency of symbols referring to containers and vice versa). Ready for review.

arporter commented:
Thanks for spotting that. I've worked through eg4 and updated the various READMEs etc. appropriately. Using `KernelGlobalsToArgs` in anger for the first time I found #663 and we also need to do #490.
Just waiting to see what Travis says now...

arporter commented:
I've had to manually check coverage but I think it's all fine. Ready for next review.

rupertford commented on src/psyclone/tests/psyir/symbols/containersymbol_test.py:
```diff
@@ -73,10 +78,12 @@ def test_containersymbol_initialisation():
     assert isinstance(sym, ContainerSymbol)
     assert sym.name == "my_mod"
-    assert not sym._reference  # Reference are not evaluated until told
+    assert not sym._reference  # References are not evaluated until told
```
As before please reword as it is not clear what the comment means.

arporter replied:
Done.

rupertford commented on src/psyclone/tests/test_files/gocean1p0/kernel_with_unqualified_use_mod.f90:
```diff
@@ -0,0 +1,92 @@
+    ! Locals
+    real(go_wp) :: amp_tide, omega_tide, rtime
+
+    amp_tide   = 0.2_go_wp
```
Is there any need for code in the body of this subroutine?

arporter replied:
I'd rather keep it as is so that it's valid Fortran which requires the import.

rupertford replied:
It's still longer than needed but I'll let it go :-)

rupertford commented on src/psyclone/transformations.py:
```diff
@@ -3810,9 +3816,14 @@ def validate(self, node, options=None):
         '''
         from psyclone.psyGen import CodedKern
         from psyclone.gocean1p0 import GOInvokeSchedule
+        from psyclone.psyir.symbols import SymbolError
```
This is imported twice. Could you import it at the start of the file (assuming it does not cause circular dependencies).

arporter replied:
The earlier import had disappeared when I came to look at it but I've moved this one to the start of the file now.

rupertford commented on src/psyclone/transformations.py:
```diff
@@ -3836,24 +3866,23 @@ def apply(self, node, options=None):
-                  transformation.
-        :rtype: (:py:class:`psyclone.psyGen.Schedule`, \
-                 :py:class:`psyclone.undoredo.Memento`).
+        :returns: a record of the transformation.
```
Returns the schedule too.

arporter replied:
Not any more (#595 )

rupertford replied:
That issue raises a question that hasn't been agreed!

arporter replied:
@hiker and I agree :-)

rupertford replied:
Well I think I agree too and we could also throw away memento given that it is not used and that the same functionality could be supported externally to transformations (but less efficiently) by copying the full PSyIR tree before each transformation. However, Sergi and Iva should also be asked for their input before making a decision, the question should be changed into a standard issue and the issue should be referenced here so people know why this transformation is different to the rest.

arporter replied:
OK, I've asked Sergi and Iva so we'll have to wait and see. Alternatively, being the impatient person that I am I could put it back to the way it used to be and then we change all transformations once we have agreed.

rupertford replied:
A separation of concerns - it could catch on :-)

rupertford commented on src/psyclone/transformations.py:
```diff
@@ -3879,3 +3912,12 @@ def apply(self, node, options=None):
 
             # Add the global variable in the call argument list
             node.arguments.append(globalvar.name)
```
Not for this PR but should node.arguments be symbols rather than strings?

arporter replied:
Possibly although that then moves us up into the "PSy layer" PSyIR and I don't think we use a symbol table there yet...

rupertford commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -195,6 +195,42 @@ def __contains__(self, key):
         '''
         return key in self._symbols
 
+    def remove(self, key):
```
Could the key be the symbol itself rather than the name, or do we need it to be the name>

arporter replied:
It could and now it is!

arporter commented on src/psyclone/tests/test_files/gocean1p0/kernel_with_unqualified_use_mod.f90:
```diff
@@ -0,0 +1,92 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 20120, Science and Technology Facilities Council
```
Oops.

arporter commented on doc/user_guide/examples.rst:
```diff
@@ -94,7 +94,14 @@ Example 4: Kernels containing use statements
+nested ``use`` of modules (i.e. data accessed via a module that in turn
+imports that symbol from another module) and kernels that call other
+kernels (Issue #342). In addition, the transformation that adds
+``!$ACC ROUTINE`` to kernel code still works with the fparser2 parse tree
```
Yes, that's better, thanks.

arporter commented on examples/gocean/eg4/README.md:
```diff
@@ -41,14 +41,16 @@ psyclone -api "gocean1.0" -s ./acc_transform.py alg.f90
+combination with the KernelGlobalsToArguments transformation (which is
+required because `kern_use_var` accesses `data_mod::gravity`.
+
+In addition, although the kernels named in the Invoke would be
```
Oh yes, thanks.

arporter commented on examples/gocean/eg4/README.md:
```diff
@@ -41,14 +41,16 @@ psyclone -api "gocean1.0" -s ./acc_transform.py alg.f90
-then call have not and this means that they won't be compiled for
-execution on the OpenACC device. Support for recursive kernel
-transformation is the subject of [Issue #342](https://github.com/stfc/PSyclone/issues/342).
+ACCRoutineTrans still works with the fparser2 parse tree and not the
```
Oh yes, thanks. That was a pain but I've hopefully sorted it now. I've broken the example into three and documented each. I've also generalised the error text so it doesn't assume that a global symbol represents data.

arporter commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -239,26 +240,64 @@ class FortranWriter(PSyIRVisitor):
+        Fortran use statement(s) for this ContainerSymbol. If this symbol
+        has both a wildcard import and explicit imports then two use
+        statements are generated. (This means that when generating Fortran
+        from a PSyIR of existing Fortran code we replicate the structure of
```
Done.

arporter commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -545,17 +545,73 @@ def process_declarations(self, parent, nodes, arg_list):
-            # Add the module symbol in the symbol table
-            container = ContainerSymbol(mod_name)
-            parent.symbol_table.add(container)
+            # Add the module symbol in the symbol table. Keep a record of
```
Yes.

arporter commented on src/psyclone/psyir/transformations/profile_trans.py:
```diff
@@ -162,7 +162,7 @@ def apply(self, nodes, options=None):
         # single node.
         from psyclone.psyGen import Node, OMPDoDirective, ACCLoopDirective
-        if isinstance(nodes, list) and isinstance(nodes[0], Node):
+        if isinstance(nodes, list) and list and isinstance(nodes[0], Node):
```
Oops!

arporter commented on src/psyclone/psyir/symbols/containersymbol.py:
```diff
@@ -85,6 +90,74 @@ def __str__(self):
+        return list(self._datasymbols)
+
+    @property
+    def has_wildcard_import(self):
```
Done.

arporter commented on src/psyclone/tests/domain/gocean/transformations/globalstoargs_test.py:
```diff
@@ -33,14 +33,16 @@
 # Author: S. Siso, STFC Daresbury Lab
 
-'''Tests the GlobalsToArgumentsTransformation for the  GOcean 1.0 API.'''
+''' Tests the GlobalsToArgumentsTransformation for the GOcean 1.0 API.'''
```
Done.

arporter commented on src/psyclone/tests/psyir/frontend/fortran_use_test.py:
```diff
@@ -0,0 +1,245 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2019, Science and Technology Facilities Council.
```
Done.

arporter commented on src/psyclone/tests/psyir/backend/fortran_test.py:
```diff
@@ -242,23 +242,69 @@ def test_gen_datatype_error(monkeypatch):
     '''Check the FortranWriter class gen_use method produces the expected
     declaration. Also check that an exception is raised if the symbol
-    does not describe a use statement.
+    does not describe a Container.
```
Has been done somewhere along the way.

arporter commented on src/psyclone/tests/psyir/frontend/fortran_use_test.py:
```diff
@@ -0,0 +1,245 @@
+        container = symtab.lookup(module_name)
+        assert isinstance(container, ContainerSymbol)
+        assert container.name == module_name
+        assert not container._reference  # It is not evaluated explicitly told
```
Weird. Corrected now.

arporter commented on src/psyclone/psyir/symbols/containersymbol.py:
```diff
@@ -85,6 +90,74 @@ def __str__(self):
+                           "'{1}'".format(data_symbol.name, self.name))
+        self._datasymbols.remove(data_symbol)
+
+    def add_wildcard_import(self):
```
Done.

arporter commented:
I've addressed all comments and updated the PSyIR 'create' example to add the import of a symbol.
Assuming Travis is happy then this is ready for next review.

arporter commented:
Now getting one failure in globalstoargs_test.py. I need to investigate...

arporter commented:
Failing test was monkeypatching Reference.check_declared() whereas now that checking is done in Node.find_symbol(). This then made me realise that we no longer need Reference.check_declared() so I've removed it and the associated test.

arporter commented:
Travis permitting, this is ready for review again.

rupertford commented on src/psyclone/psyir/nodes/reference.py:
```diff
@@ -120,33 +120,6 @@ def reference_accesses(self, var_accesses):
         '''
         var_accesses.add_access(self.name, AccessType.READ, self)
 
```
Could you please explain how removing check_declared is related to this PR. Sorry if I'm slow on the uptake.

arporter replied:
Strictly speaking it's not - I just realised that it is dead code while updating the tests.

rupertford commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -579,17 +579,74 @@ def process_declarations(self, parent, nodes, arg_list):
+                    # will replace a previous import with an empty only-list.
+                    pass
+                container.wildcard_import = True
+            elif decl.items[3].lower().replace(" ", "") == ",only:":
```
Could this be `if decl.items[3] and not decl.items[4]` i.e. has only clause but does not have any arguments to it?

arporter replied:
We are already within a `if not isinstance(decl.items[4], Fortran2003.Only_List)` so we know there's no list. I could then just check that decl.items[3] is not empty which would just be a bare `else` on the existing if-block but I know you don't like that. This way we get to check that `decl.items[3]` is what we expect but I'm happy to change if you want.

rupertford replied:
No, you are right. Looking back at the code this is clear.

rupertford commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -579,17 +579,74 @@ def process_declarations(self, parent, nodes, arg_list):
+            elif decl.items[3].lower().replace(" ", "") == ",only:":
+                # This use has an 'only: ' but no associated list of
+                # imported symbols. (It serves to keep a module in scope while
+                # not actually importing anything from it.)
```
Please explain that nothing needs to be set as an empty only list (default) and no wildcard import (default) implies `only:`

arporter replied:
Done.

rupertford commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -579,17 +579,74 @@ def process_declarations(self, parent, nodes, arg_list):
+                   (container.wildcard_import or
+                    parent.symbol_table.imported_symbols(container)):
+                    # TODO #11 Log the fact that this import with an empty
+                    # only-list is irrelevant because of existing 'use's of
```
Perhaps change `irrelevant` to `ignored` as an existing container remains unchanged if `only:` is found.

arporter replied:
Done.

rupertford commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -579,17 +579,74 @@ def process_declarations(self, parent, nodes, arg_list):
+                if (not new_container) and (not container.wildcard_import) \
+                   and (not parent.symbol_table.imported_symbols(container)):
+                    # TODO #11 Log the fact that this explicit symbol import
+                    # will replace a previous import with an empty only-list.
```
Please make it clearer that "a previous import with an empty only-list" is being replaced, rather than "a previous import" being replaced with "an empty only-list".

arporter replied:
Hopefully done.

rupertford commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -136,6 +137,79 @@ def test_add():
+
+def test_remove():
+    '''Test that the remove method removes symbols from the symbol
+    table, but raises appropiate errors when provided with wrong
```
Update text to say that it raises and exception if symbols exist in the container. Also point out this is just testing container symbols at the moment.

arporter replied:
Good point. Done.

rupertford replied:
There is also a small typo "appropiate"

arporter replied:
Done.

rupertford commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -136,6 +137,79 @@ def test_add():
+        sym_table.remove(my_mod)
+    assert ("Cannot remove Symbol 'my_mod' from symbol table because it does "
+            "not" in str(err.value))
+    # Attempt to supply something that is not a Symbol
```
You could add a test to try to remove a symbol that is in the symbol table but is not a container symbol.

arporter replied:
Good idea. Done.

arporter commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -196,6 +196,68 @@ def __contains__(self, key):
+        :raises KeyError: if the supplied symbol is not in the symbol table.
+        :raises ValueError: if the supplied container symbol is referenced \
+                            by one or more DataSymbols.
+        :raises TypeError: if the supplied symbol is not a ContainerSymbol.
```
Done.

arporter commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -196,6 +196,68 @@ def __contains__(self, key):
+
+    def remove(self, symbol):
+        ''' Remove the supplied ContainerSymbol from the Symbol Table.
+
```
Yes, I wondered about that. Done.

arporter commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -344,6 +406,15 @@ def precision_datasymbols(self):
         return list(precision_symbols)
 
+    @property
+    def container_symbols(self):
```
I suppose so :-) I've removed the underscore here as we use "datasymbol" a lot.

arporter commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -196,6 +196,68 @@ def __contains__(self, key):
+        :raises TypeError: if the supplied object is not a ContainerSymbol.
+
+        '''
+        if not isinstance(csymbol, ContainerSymbol):
```
Oh yes, good point. I moved this method into the SymbolTable and forgot this check.

rupertford replied:
As discussed, could you please add a comment explaining that a KeyError is raised if the container symbol is not in the symbol table.

arporter replied:
Done.

arporter commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -196,6 +196,68 @@ def __contains__(self, key):
+                "imported_symbols() expects a ContainerSymbol but got an "
+                "object of type '{0}'".format(type(csymbol).__name__))
+        symbol_list = []
+        for symbol in self.global_datasymbols:
```
It could and now it is.

rupertford replied:
Nice :-)

arporter commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -398,18 +469,21 @@ def copy_external_global(self, globalvar):
             new_symbol = globalvar.copy()
-            container_ref = self.lookup(external_container_name)
+            # Update the interface of this new symbol. This also removes the
+            # symbol from its original Container and adds it to the new one.
```
Oops, no. That's what used to happen. I've removed it.

arporter commented on src/psyclone/psyir/symbols/containersymbol.py:
```diff
@@ -85,6 +88,31 @@ def __str__(self):
+    @property
+    def wildcard_import(self):
+        '''
+        :returns: whether or not there is a wildcard import of all public
```
Oops.

arporter commented on src/psyclone/psyir/symbols/containersymbol.py:
```diff
@@ -85,6 +88,31 @@ def __str__(self):
+        :returns: whether or not there is a wildcard import of all public
+                  symbols from this Container.
+        :rtype: bool
+        '''
```
Done.

arporter commented on src/psyclone/psyir/symbols/containersymbol.py:
```diff
@@ -85,6 +88,31 @@ def __str__(self):
+    @wildcard_import.setter
+    def wildcard_import(self, value):
+        '''
+        Set whether or not there is a wildcard import of all public symbols
```
A backslash is only required for the :param:/:returns:/:rtype: etc. descriptions

rupertford replied:
Oops :-)

arporter commented on examples/gocean/eg4/README.md:
````diff
@@ -13,50 +13,71 @@ PSyclone can be run in the directory containing this file by
 
 ```sh
-psyclone -api "gocean1.0" alg.f90
+psyclone -api "gocean1.0" alg_kern_with_use.f90
````
We can do OpenCL so I've added that (plus the extra line in the check_examples script).

arporter commented on examples/psyir/create.py:
```diff
@@ -49,8 +49,8 @@
 from psyclone.psyGen import KernelSchedule
-from psyclone.psyir.symbols import DataSymbol, SymbolTable, \
-    ArgumentInterface, DataType
+from psyclone.psyir.symbols import DataSymbol, SymbolTable, ContainerSymbol, \
```
Done.

arporter replied:
Done.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2649,16 +2649,14 @@ def validate(self, sched, options=None):
                     "The Symbol Table for kernel '{0}' contains the following "
-                    "symbols with 'global' scope: {1}. PSyclone cannot "
-                    "currently transform such a kernel into OpenCL.".
+                    "symbols with 'global' scope: {1}. All data accessed by "
```
Oh yes. Re-worded to cover this case.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -3306,21 +3307,25 @@ def validate(self, kern, options=None):
-                "symbols with 'global' scope: {1}. PSyclone cannot currently "
-                "transform kernels for execution on an OpenACC device if "
-                "they access data not passed by argument.".
+                "symbol(s) with global scope: {1}. PSyclone cannot currently "
```
I've updated the error message. As it happens, the two cases are different since I think we simply can't call another kernel from within an OpenCL kernel whereas we have a ticket to support this for OpenACC.

arporter commented:
Assuming Travis is happy then this is ready for next review.

rupertford commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -579,17 +579,77 @@ def process_declarations(self, parent, nodes, arg_list):
+                # wildcard import) imply 'only:'.
+                if not new_container and \
+                   (container.wildcard_import or
+                    parent.symbol_table.imported_symbols(container)):
```
E129 visually indented line with same indent as next logical line

arporter replied:
I hate that error! However, it has forced me to google and actually, the recommended solution is OK.

rupertford commented on src/psyclone/psyir/nodes/reference.py:
```diff
@@ -120,33 +120,6 @@ def reference_accesses(self, var_accesses):
-        # issue #500. After this change has been made this method could
-        # make use of the get_symbol method to determine
-        # whether the reference has been declared (or not).
-        if found_symbol_table:
```
W: 44, 0: Unused SymbolError imported from psyclone.psyir.symbols (unused-import)

arporter replied:
Fixed.

rupertford commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -196,6 +196,72 @@ def __contains__(self, key):
+                    [sym.name for sym in self.imported_symbols(symbol)]))
+        self._symbols.pop(symbol.name)
+
+
```
symboltable.py:265:5: E303 too many blank lines (2)

arporter replied:
Fixed.

rupertford commented on src/psyclone/tests/psyir/nodes/reference_test.py:
```diff
@@ -195,21 +195,3 @@ def test_array_create_invalid3():
         "should be a PSyIR Node but found 'str'." in str(excinfo.value))
-
-
-def test_reference_check_declared():
```
W: 44, 0: Unused SymbolTable imported from psyclone.psyir.symbols (unused-import)

arporter replied:
Fixed.

rupertford commented on src/psyclone/tests/psyir/symbols/containersymbol_test.py:
```diff
@@ -41,9 +41,12 @@
-from psyclone.psyir.symbols import ContainerSymbol, SymbolError
-from psyclone.psyir.symbols.containersymbol import ContainerSymbolInterface, \
-    FortranModuleInterface
+from psyclone.psyir.symbols import SymbolError
```
W: 47, 0: Unused LocalInterface imported from psyclone.psyir.symbols.datasymbol (unused-import)
W: 47, 0: Unused DataSymbol imported from psyclone.psyir.symbols.datasymbol (unused-import)
W: 47, 0: Unused GlobalInterface imported from psyclone.psyir.symbols.datasymbol (unused-import)
W: 49, 0: Unused DataType imported from psyclone.psyir.symbols.datatypes (unused-import)

arporter replied:
Oops. Fixed.

arporter commented:
Ready for review again.
